{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "526a5f8f",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“package ‘devtools’ was built under R version 4.3.3”\n"
     ]
    }
   ],
   "source": [
    "suppressMessages(library(ggplot2))\n",
    "suppressMessages(library(ArchR))\n",
    "suppressMessages(library(Seurat))\n",
    "suppressMessages(library(Signac))\n",
    "suppressMessages(library(patchwork))\n",
    "suppressMessages(library(dplyr))\n",
    "suppressMessages(library(cowplot))\n",
    "suppressMessages(library(ggrepel))\n",
    "suppressMessages(library(rhdf5))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "88575b42",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting default number of Parallel threads to 64.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "set.seed(42)\n",
    "addArchRThreads(threads = 64)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4b7d34eb",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Successfully loaded ArchRProject!\n",
      "\n",
      "\n",
      "           ___      .______        ______  __    __  .______      \n",
      "          /   \\     |   _  \\      /      ||  |  |  | |   _  \\     \n",
      "         /  ^  \\    |  |_)  |    |  ,----'|  |__|  | |  |_)  |    \n",
      "        /  /_\\  \\   |      /     |  |     |   __   | |      /     \n",
      "       /  _____  \\  |  |\\  \\\\___ |  `----.|  |  |  | |  |\\  \\\\___.\n",
      "      /__/     \\__\\ | _| `._____| \\______||__|  |__| | _| `._____|\n",
      "    \n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "class: ArchRProject \n",
       "outputDirectory: /home/xwt/27/BCY/BCY-pair/data_integration/fragments/ArchRProject \n",
       "samples(6): 10N 10T ... 12N 22T\n",
       "sampleColData names(1): ArrowFiles\n",
       "cellColData names(19): Sample TSSEnrichment ... Clusters\n",
       "  predictedCellType\n",
       "numberOfCells(1): 49601\n",
       "medianTSS(1): 17.261\n",
       "medianFrags(1): 15163"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "proj <- loadArchRProject(\"./fragments/ArchRProject\", showLogo = FALSE)\n",
    "proj"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "obj.atac.annotated <- readRDS(\"./fragments/ArchRProject/scATAC.annotated.Rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ee0cc635",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "new_order <- c(\"T cells\", \"B cells\", \"Myeloid\", \"Mast\", \"Endothelial\", \"Plasmablast\", \"Epithelial\", \"Fibroblast\", \"PeriVascular\")\n",
    "obj.atac.annotated$majorType <- factor(obj.atac.annotated$majorType, levels = new_order)\n",
    "cols <- ArchR::paletteDiscrete(obj.atac.annotated$majorType)\n",
    "options(repr.plot.width = 7, repr.plot.height = 6)\n",
    "p3 <- DimPlot(object = obj.atac.annotated, reduction = \"umap_harmony\",\n",
    "              group.by = \"majorType\", shuffle = TRUE) +\n",
    "              scale_color_manual(values = cols) +\n",
    "              xlab(\"UMAP1\") + ylab(\"UMAP2\") + \n",
    "              ggtitle(\"\") +\n",
    "              theme(axis.title = element_blank(), axis.line = element_blank(), axis.ticks = element_blank(), axis.text = element_blank())\n",
    "\n",
    "ggsave(\"./output/CellType_UMAP.svg\", plot = p3, width = 7, height = 6, device = \"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "6"
      ],
      "text/latex": [
       "6"
      ],
      "text/markdown": [
       "6"
      ],
      "text/plain": [
       "[1] 6"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'Sample'. You can override using the\n",
      "`.groups` argument.\n"
     ]
    }
   ],
   "source": [
    "df <- as.data.frame(obj.atac.annotated@meta.data)\n",
    "length(unique(df$Sample))\n",
    "\n",
    "df_ct <- df %>%\n",
    "    group_by(Sample, majorType) %>%\n",
    "    summarise(counts = n()) %>%\n",
    "    mutate(cell_proportion = counts / sum(counts))\n",
    "\n",
    "df_ct$majorType <- factor(df_ct$majorType, levels = new_order)\n",
    "cols <- ArchR::paletteDiscrete(df_ct$majorType)\n",
    "\n",
    "p2 <- ggplot(df_ct, aes(Sample, cell_proportion, fill=majorType)) + \n",
    "    geom_bar(stat=\"identity\", position = position_stack(reverse = TRUE)) +\n",
    "    scale_fill_manual(values = cols) +\n",
    "    theme_cowplot() +\n",
    "    xlab(\"\") + ylab(\"\") +\n",
    "    theme(axis.text.x = element_text(angle = 60, hjust = 1)) +\n",
    "    scale_y_continuous(labels = scales::percent_format(scale = 100), \n",
    "                       expand = expansion(mult = c(0, 0.05))) + \n",
    "    coord_cartesian(ylim = c(0, 1))\n",
    "    \n",
    "\n",
    "options(repr.plot.width = 6, repr.plot.height = 6)\n",
    "\n",
    "ggsave(\"./output/CellType_proportion_sample.svg\", plot = p2, width = 6, height = 6, device = \"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4ae60c31",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "markersGS <- readRDS(\"./fragments/ArchRProject_clean/markersGS_predictedCellType.rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "88d848bb",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-plotMarkerHeatmap-1e41c7423248d5-Date-2025-11-09_Time-15-30-24.040861.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Printing Top Marker Genes:\n",
      "\n",
      "B cells:\n",
      "\n",
      "\tMIR4420, RNU5F-1, BEND5, MIR4422, MIR7852, SLC16A4, DRAM2, TTC24, RALGPS2, MIR29C, FAM177B, KMO, OR14I1, IDI2-AS1, USP6NL\n",
      "\n",
      "Endothelial:\n",
      "\n",
      "\tVWA1, LINC01770, RBP7, EPHA2, HSPG2, FAM110D, SLFNL1-AS1, TIE1, ADGRL4, LINC01361, ABCA4, LINC01708, PALMD, FAM72D, LCE3E\n",
      "\n",
      "Epithelial:\n",
      "\n",
      "\tRNF223, MIR200B, MIR200A, MIR429, MIR6727, TAS1R3, ATAD3C, ARHGEF16, HES2, ESPN, MIR4252, CASZ1, PADI3, KLHDC7A, AKR7L\n",
      "\n",
      "Fibroblast:\n",
      "\n",
      "\tGPR153, LINC01672, MIR34A, MIR6729, PRAMEF2, PRAMEF17, PRAMEF20, LRRC38, PDPN, PLEKHM2, PLA2G5, FAM43B, MIR4684, TCEA3, A3GALT2\n",
      "\n",
      "Mast:\n",
      "\n",
      "\tUTS2, NPPA-AS1, PLA2G2C, TRIM63, MIR6735, NFIA-AS2, MIR3117, TCTEX1D1, ALX3, ATP1A4, RGS13, LHX9, SLC45A3, SPATA45, GCSAML\n",
      "\n",
      "Myeloid:\n",
      "\n",
      "\tTNFRSF14, MIR4632, PADI2, CELA3A, MIR6127, C1QC, C1QB, FGR, PTAFR, FAM151A, DNASE2B, GNAT2, AMPD2, CTSS, RPTN\n",
      "\n",
      "PeriVascular:\n",
      "\n",
      "\tLINC00982, PRDM16, NPPB, FBLIM1, UQCRHL, GJA4, NT5C1A, CYP4X1, SGIP1, CYR61, PLPPR4, INKA2-AS1, ETV3L, OR10R2, OR6Y1\n",
      "\n",
      "Plasmablast:\n",
      "\n",
      "\tSLC2A7, CORT, PLA2G2D, LOC100506985, BMP8A, BMP8B, OXCT2, MIR4421, GPR88, AMY2B, NTNG1, KCNA2, AMPD1, TENT5C, MIR921\n",
      "\n",
      "T cells:\n",
      "\n",
      "\tTNFRSF18, TNFRSF4, MMP23A, TNFRSF9, PIK3CD-AS1, TNFRSF1B, RUNX3, MIR6731, MIR4425, ZNF683, LCK, TMCO2, SMAP2, IL23R, IL12RB2\n",
      "\n",
      "Identified 5018 markers!\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " [1] \"CD38\"   \"IGLL5\"  \"PAX5\"   \"ITGAM\"  \"FAP\"    \"FOXP3\"  \"VIM\"    \"PDCD1\" \n",
      " [9] \"CD8A\"   \"CD3D\"   \"CD4\"    \"CXCL12\" \"THY1\"   \"PECAM1\" \"KDR\"    \"EPACM\" \n",
      "[17] \"ERBB2\"  \"FOXA1\"  \"KRT8\"   \"ESR1\"   \"KRT14\" \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Preparing Main Heatmap..\n",
      "\n",
      "'magick' package is suggested to install to give better rasterization.\n",
      "\n",
      "Set `ht_opt$message = FALSE` to turn off this message.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-plotMarkerHeatmap-1e41c7423248d5-Date-2025-11-09_Time-15-30-24.040861.log\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>agg_record_180837382:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{agg\\textbackslash{}\\_record\\textbackslash{}\\_180837382:} 2"
      ],
      "text/markdown": [
       "**agg_record_180837382:** 2"
      ],
      "text/plain": [
       "agg_record_180837382 \n",
       "                   2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "markerGenes <- c(\n",
    "    \"CD38\", \"IGLL5\", \"PAX5\", \"ITGAM\", \"FAP\",\n",
    "    \"FOXP3\", \"VIM\", \"PDCD1\", \"CD8A\", \"CD3D\",\n",
    "    \"CD4\", \"CXCL12\", \"THY1\", \"PECAM1\", \"KDR\",\n",
    "    \"EPACM\", \"ERBB2\", \"FOXA1\", \"KRT8\", \"ESR1\",\n",
    "    \"KRT14\"\n",
    ")\n",
    "\n",
    "heatmapGS <- plotMarkerHeatmap(\n",
    "  seMarker = markersGS, \n",
    "  cutOff = \"FDR <= 0.05 & Log2FC >= 1.25\", \n",
    "  labelMarkers = markerGenes,\n",
    "  clusterCols = FALSE\n",
    ")\n",
    "\n",
    "idx <- which(mcols(markersGS)$name %in% markerGenes)\n",
    "mtx <- assays(markersGS[idx,])$Mean %>% `rownames<-`(., mcols(markersGS)$name[idx])\n",
    "mtx <- t(scale(t(mtx)))\n",
    "fh <- function(x) hclust(dist(x), method=\"ward.D2\")\n",
    "col_fun1 <- colorRamp2(c(-2,-1,0,1,2), paletteContinuous(set = \"horizonExtra\", n = 5))\n",
    "\n",
    "\n",
    "pdf(\"./output/heatmapGS_z_CellType.pdf\", width = 6, height = 8)\n",
    "options(repr.plot.height = 8, repr.plot.width = 6)\n",
    "ht1 <- Heatmap(mtx, \n",
    "        name = \"heatmapGS_z\",\n",
    "        cluster_rows = fh, \n",
    "        cluster_columns = F, \n",
    "        show_row_dend = F, \n",
    "        col = col_fun1,\n",
    "        row_names_gp = gpar(fontsize = 12),\n",
    "        column_names_gp = gpar(fontsize = 12),\n",
    "        heatmap_legend_param = list(\n",
    "                 title_gp = gpar(fontsize = 12,\n",
    "                 labels_gp = gpar(fontsize = 10),\n",
    "                 legend_direction = \"horizontal\", \n",
    "                 legend_width = unit(4, \"cm\"),\n",
    "                 legend_height = unit(3, \"cm\"),\n",
    "                 title_position = \"topcenter\"\n",
    "               )\n",
    "        )\n",
    "      )\n",
    "p3 <- draw(ht1, heatmap_legend_side = \"bottom\")\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c3ea7e54",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DataFrame with 6 rows and 6 columns\n",
      "    seqnames       idx                   name combinedVars combinedMeans\n",
      "       <Rle> <integer>            <character>    <numeric>     <numeric>\n",
      "f90        z        90 Fox.Ebox.Forkhead.bH..      67.6093     -0.315460\n",
      "f87        z        87      FOXA1.Forkhead_87      64.6130     -0.331883\n",
      "f1         z         1            AP.1.bZIP_1      60.7542     -0.300232\n",
      "f86        z        86          Fosl2.bZIP_86      59.4828     -0.309931\n",
      "f94        z        94      FOXM1.Forkhead_94      59.3852     -0.299776\n",
      "f99        z        99           Fra2.bZIP_99      58.8002     -0.330543\n",
      "         rank\n",
      "    <integer>\n",
      "f90         1\n",
      "f87         2\n",
      "f1          3\n",
      "f86         4\n",
      "f94         5\n",
      "f99         6\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-getMarkerFeatures-1e41c760c972e0-Date-2025-11-09_Time-15-31-01.413643.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "MatrixClass = Sparse.Assays.Matrix\n",
      "\n",
      "2025-11-09 15:31:01.585875 : Matching Known Biases, 0.002 mins elapsed.\n",
      "\n",
      "Found less than 100 cells for background matching, Lowering k to 0\n",
      "\n",
      "2025-11-09 15:31:04.470547 : \n",
      "\n"
     ]
    }
   ],
   "source": [
    "## motif analysis - Cell Type\n",
    "VarMotifs <- getVarDeviations(proj, name = \"homerMatrix\", plot = F)\n",
    "MotifScoreClusters <- getMarkerFeatures(proj, \n",
    "                                        useMatrix = \"homerMatrix\", \n",
    "                                        groupBy = \"predictedCellType\",\n",
    "                                        bias = c(\"TSSEnrichment\", \"log10(nFrags)\"),\n",
    "                                        testMethod = \"wilcoxon\", \n",
    "                                        useSeqnames = \"z\")\n",
    "fh <- function(x) hclust(dist(x), method=\"ward.D2\")\n",
    "idy <- which(mcols(MotifScoreClusters)$name %in% VarMotifs$name[c(1:50)])\n",
    "mtx2 <- assays(MotifScoreClusters[idy,])$Mean %>% `rownames<-`(., mcols(MotifScoreClusters)$name[idy])\n",
    "col_fun2 <- colorRamp2(c(-5,-2.5,0,2,5), paletteContinuous(set = \"solarExtra\", n = 5))\n",
    "pdf(\"./output/heatmap_Motif_CellType.pdf\", width = 8, height = 8)\n",
    "ht2 <- Heatmap(mtx2, \n",
    "        name = \"Mean motif score\", \n",
    "        cluster_rows = fh, \n",
    "        cluster_columns = F, \n",
    "        show_row_dend = F, \n",
    "        col = col_fun2,\n",
    "        row_names_gp = gpar(fontsize = 12),\n",
    "        column_names_gp = gpar(fontsize = 12),\n",
    "        heatmap_legend_param = list(\n",
    "                 title_gp = gpar(fontsize = 12),\n",
    "                 labels_gp = gpar(fontsize = 10), \n",
    "                 legend_direction = \"horizontal\",\n",
    "                 legend_width = unit(4, \"cm\"),\n",
    "                 legend_height = unit(3, \"cm\"),\n",
    "                 title_position = \"topcenter\"\n",
    "               )\n",
    "        )\n",
    "\n",
    "p6 <- draw(ht2, heatmap_legend_side = \"bottom\")\n",
    "p6\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a6242c0c",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class: SummarizedExperiment \n",
       "dim: 298721 9 \n",
       "metadata(2): MatchInfo Params\n",
       "assays(7): Log2FC Mean ... AUC MeanBGD\n",
       "rownames(298721): 1 2 ... 298720 298721\n",
       "rowData names(4): seqnames idx start end\n",
       "colnames(9): B cells Endothelial ... Plasmablast T cells\n",
       "colData names(0):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## motif analysis - cell types motif enrichment\n",
    "PeakCellType <- readRDS(\"./fragments/ArchRProject_clean/markersPK_predictedCellType.rds\")\n",
    "PeakCellType"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "390ba63e",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-plotMarkerHeatmap-1e41c72cbeb47-Date-2025-11-09_Time-15-31-17.738615.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Identified 153244 markers!\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "character(0)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Adding Annotations..\n",
      "\n",
      "Preparing Main Heatmap..\n",
      "\n",
      "'magick' package is suggested to install to give better rasterization.\n",
      "\n",
      "Set `ht_opt$message = FALSE` to turn off this message.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-plotMarkerHeatmap-1e41c72cbeb47-Date-2025-11-09_Time-15-31-17.738615.log\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>agg_record_1120477757:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{agg\\textbackslash{}\\_record\\textbackslash{}\\_1120477757:} 2"
      ],
      "text/markdown": [
       "**agg_record_1120477757:** 2"
      ],
      "text/plain": [
       "agg_record_1120477757 \n",
       "                    2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "heatmapPeaks <- plotMarkerHeatmap(\n",
    "    seMarker = PeakCellType,\n",
    "    cutOff = \"FDR < 0.01 & Log2FC >= 1\",\n",
    "    transpose = TRUE,\n",
    "    clusterCols = FALSE,\n",
    "    nLabel = 0\n",
    ")\n",
    "options(repr.plot.height = 10, repr.plot.width = 10)\n",
    "\n",
    "pdf(\"./output/heatmapPK_CellType.pdf\", width = 10, height = 10)\n",
    "ComplexHeatmap::draw(heatmapPeaks, \n",
    "                     heatmap_legend_side = \"bot\", \n",
    "                     annotation_legend_side = \"bot\")\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "3ae9717b",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-peakAnnoEnrichment-1e41c71b68af92-Date-2025-11-09_Time-15-31-34.217354.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-11-09 15:31:45.598196 : Computing Enrichments 1 of 9, 0.19 mins elapsed.\n",
      "\n",
      "2025-11-09 15:31:45.800749 : Computing Enrichments 2 of 9, 0.193 mins elapsed.\n",
      "\n",
      "2025-11-09 15:31:45.956959 : Computing Enrichments 3 of 9, 0.196 mins elapsed.\n",
      "\n",
      "2025-11-09 15:31:46.134739 : Computing Enrichments 4 of 9, 0.199 mins elapsed.\n",
      "\n",
      "2025-11-09 15:31:46.30435 : Computing Enrichments 5 of 9, 0.201 mins elapsed.\n",
      "\n",
      "2025-11-09 15:31:46.448942 : Computing Enrichments 6 of 9, 0.204 mins elapsed.\n",
      "\n",
      "2025-11-09 15:31:46.608091 : Computing Enrichments 7 of 9, 0.207 mins elapsed.\n",
      "\n",
      "2025-11-09 15:31:46.76103 : Computing Enrichments 8 of 9, 0.209 mins elapsed.\n",
      "\n",
      "2025-11-09 15:31:46.924539 : Computing Enrichments 9 of 9, 0.212 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-peakAnnoEnrichment-1e41c71b68af92-Date-2025-11-09_Time-15-31-34.217354.log\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "class: SummarizedExperiment \n",
       "dim: 332 9 \n",
       "metadata(0):\n",
       "assays(10): mlog10Padj mlog10p ... CompareFrequency feature\n",
       "rownames(332): AP.1.bZIP_1 AP.2gamma.AP2_2 ... ZNF711.Zf_331\n",
       "  ZSCAN22.Zf_332\n",
       "rowData names(0):\n",
       "colnames(9): B cells Endothelial ... Plasmablast T cells\n",
       "colData names(0):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-plotEnrichHeatmap-1e41c774af163e-Date-2025-11-09_Time-15-31-47.204026.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "Adding Annotations..\n",
      "\n",
      "Preparing Main Heatmap..\n",
      "\n",
      "'magick' package is suggested to install to give better rasterization.\n",
      "\n",
      "Set `ht_opt$message = FALSE` to turn off this message.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>agg_record_1726491565:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{agg\\textbackslash{}\\_record\\textbackslash{}\\_1726491565:} 2"
      ],
      "text/markdown": [
       "**agg_record_1726491565:** 2"
      ],
      "text/plain": [
       "agg_record_1726491565 \n",
       "                    2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "enrichMotifs <- ArchR::peakAnnoEnrichment(\n",
    "  seMarker = PeakCellType,\n",
    "  ArchRProj = proj,\n",
    "  peakAnnotation = \"homer\",\n",
    "  cutOff = \"FDR < 0.01 & Log2FC >= 1\"\n",
    ")\n",
    "enrichMotifs\n",
    "heatmapMotif <- plotEnrichHeatmap(enrichMotifs, n = 7, clusterCols = F, transpose = TRUE)\n",
    "options(repr.plot.height = 10, repr.plot.width = 10)\n",
    "pdf(\"./output/heatmapPK_motif_CellType.pdf\", width = 10, height = 10)\n",
    "ComplexHeatmap::draw(heatmapMotif, \n",
    "                     heatmap_legend_side = \"bot\", \n",
    "                     annotation_legend_side = \"bot\")\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "0487f14b",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-plotBrowserTrack-1e41c734fe364e-Date-2025-11-09_Time-15-31-49.924781.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2025-11-09 15:31:50.001576 : Validating Region, 0.001 mins elapsed.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GRanges object with 5 ranges and 2 metadata columns:\n",
      "      seqnames            ranges strand |     gene_id      symbol\n",
      "         <Rle>         <IRanges>  <Rle> | <character> <character>\n",
      "  [1]    chr21 45073853-45226560      + |         104      ADARB1\n",
      "  [2]    chr21 45263928-45287898      - |       23275      POFUT2\n",
      "  [3]    chr17 59109951-59155269      - |      348235        SKA2\n",
      "  [4]    chr17 59331689-59401729      + |      388403       YPEL2\n",
      "  [5]    chr16 68737225-68835548      + |         999        CDH1\n",
      "  -------\n",
      "  seqinfo: 24 sequences from hg38 genome\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-11-09 15:31:50.07093 : Adding Bulk Tracks (1 of 5), 0.002 mins elapsed.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "R_zmq_msg_send errno: 4 strerror: Interrupted system call\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-11-09 15:31:57.05232 : Adding Feature Tracks (1 of 5), 0.119 mins elapsed.\n",
      "\n",
      "2025-11-09 15:31:57.157979 : Adding Gene Tracks (1 of 5), 0.121 mins elapsed.\n",
      "\n",
      "2025-11-09 15:31:57.499059 : Plotting, 0.126 mins elapsed.\n",
      "\n",
      "2025-11-09 15:31:58.614536 : Adding Bulk Tracks (2 of 5), 0.145 mins elapsed.\n",
      "\n",
      "2025-11-09 15:32:00.04298 : Adding Feature Tracks (2 of 5), 0.169 mins elapsed.\n",
      "\n",
      "2025-11-09 15:32:00.11223 : Adding Gene Tracks (2 of 5), 0.17 mins elapsed.\n",
      "\n",
      "2025-11-09 15:32:00.479774 : Plotting, 0.176 mins elapsed.\n",
      "\n",
      "2025-11-09 15:32:01.389142 : Adding Bulk Tracks (3 of 5), 0.191 mins elapsed.\n",
      "\n",
      "2025-11-09 15:32:04.986119 : Adding Feature Tracks (3 of 5), 0.251 mins elapsed.\n",
      "\n",
      "2025-11-09 15:32:05.069726 : Adding Gene Tracks (3 of 5), 0.252 mins elapsed.\n",
      "\n",
      "2025-11-09 15:32:05.399436 : Plotting, 0.258 mins elapsed.\n",
      "\n",
      "2025-11-09 15:32:06.277761 : Adding Bulk Tracks (4 of 5), 0.273 mins elapsed.\n",
      "\n",
      "2025-11-09 15:32:08.483752 : Adding Feature Tracks (4 of 5), 0.309 mins elapsed.\n",
      "\n",
      "2025-11-09 15:32:08.557696 : Adding Gene Tracks (4 of 5), 0.311 mins elapsed.\n",
      "\n",
      "2025-11-09 15:32:08.831707 : Plotting, 0.315 mins elapsed.\n",
      "\n",
      "2025-11-09 15:32:09.982282 : Adding Bulk Tracks (5 of 5), 0.334 mins elapsed.\n",
      "\n",
      "2025-11-09 15:32:12.26637 : Adding Feature Tracks (5 of 5), 0.372 mins elapsed.\n",
      "\n",
      "2025-11-09 15:32:12.335529 : Adding Gene Tracks (5 of 5), 0.374 mins elapsed.\n",
      "\n",
      "2025-11-09 15:32:12.63576 : Plotting, 0.379 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-plotBrowserTrack-1e41c734fe364e-Date-2025-11-09_Time-15-31-49.924781.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "markerGenes <- c(\n",
    "    \"ADARB1\", \n",
    "    \"POFUT2\", \n",
    "    \"SKA2\", \n",
    "    \"YPEL2\",\n",
    "    \"CDH1\"\n",
    ")\n",
    "\n",
    "p <- plotBrowserTrack(\n",
    "    ArchRProj = proj, \n",
    "    groupBy = \"Sample\", \n",
    "    geneSymbol = markerGenes, \n",
    "    upstream = 50000,\n",
    "    downstream = 50000\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "b717010a",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Plotting Gtable!\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NULL\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Plotting Gtable!\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NULL\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Plotting Gtable!\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NULL\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Plotting Gtable!\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NULL\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Plotting Gtable!\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NULL\n"
     ]
    }
   ],
   "source": [
    "plotPDF(plotList = p, \n",
    "    name = \"Plot-Tracks-Marker-Genes.pdf\", \n",
    "    ArchRProj = proj, \n",
    "    addDOC = FALSE, width = 5, height = 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "60ef2de0",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.1 (2023-06-16)\n",
       "Platform: x86_64-conda-linux-gnu (64-bit)\n",
       "Running under: Ubuntu 22.04.2 LTS\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /home/xwt/miniconda3/envs/YBC_NAT/lib/libopenblasp-r0.3.30.so;  LAPACK version 3.12.0\n",
       "\n",
       "Random number generation:\n",
       " RNG:     L'Ecuyer-CMRG \n",
       " Normal:  Inversion \n",
       " Sample:  Rejection \n",
       " \n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Etc/UTC\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       " [1] parallel  stats4    grid      stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] nabor_0.5.0                 circlize_0.4.16            \n",
       " [3] ComplexHeatmap_2.10.0       ggrepel_0.9.6              \n",
       " [5] cowplot_1.2.0               dplyr_1.1.4                \n",
       " [7] patchwork_1.3.2             Signac_1.14.0              \n",
       " [9] Seurat_5.0.1                SeuratObject_5.2.0         \n",
       "[11] sp_2.2-0                    rhdf5_2.46.1               \n",
       "[13] SummarizedExperiment_1.32.0 Biobase_2.62.0             \n",
       "[15] RcppArmadillo_15.0.2-2      Rcpp_1.1.0                 \n",
       "[17] Matrix_1.6-5                GenomicRanges_1.54.1       \n",
       "[19] GenomeInfoDb_1.38.8         IRanges_2.36.0             \n",
       "[21] S4Vectors_0.40.2            BiocGenerics_0.48.1        \n",
       "[23] sparseMatrixStats_1.14.0    MatrixGenerics_1.14.0      \n",
       "[25] matrixStats_1.5.0           data.table_1.17.8          \n",
       "[27] stringr_1.6.0               plyr_1.8.9                 \n",
       "[29] magrittr_2.0.4              gtable_0.3.6               \n",
       "[31] gtools_3.9.5                gridExtra_2.3              \n",
       "[33] devtools_2.4.5              usethis_3.2.1              \n",
       "[35] ArchR_1.0.3                 ggplot2_3.5.2              \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] RcppAnnoy_0.0.22            splines_4.3.1              \n",
       "  [3] later_1.4.4                 pbdZMQ_0.3-14              \n",
       "  [5] bitops_1.0-9                tibble_3.3.0               \n",
       "  [7] polyclip_1.10-7             DirichletMultinomial_1.36.0\n",
       "  [9] fastDummies_1.7.5           lifecycle_1.0.4            \n",
       " [11] doParallel_1.0.17           globals_0.18.0             \n",
       " [13] lattice_0.22-7              MASS_7.3-60                \n",
       " [15] plotly_4.11.0               remotes_2.5.0              \n",
       " [17] httpuv_1.6.16               otel_0.2.0                 \n",
       " [19] sctransform_0.4.2           spam_2.11-1                \n",
       " [21] sessioninfo_1.2.3           pkgbuild_1.4.8             \n",
       " [23] spatstat.sparse_3.1-0       reticulate_1.44.0          \n",
       " [25] DBI_1.2.3                   pbapply_1.7-4              \n",
       " [27] RColorBrewer_1.1-3          abind_1.4-8                \n",
       " [29] pkgload_1.4.1               zlibbioc_1.48.2            \n",
       " [31] Rtsne_0.17                  purrr_1.2.0                \n",
       " [33] RCurl_1.98-1.17             GenomeInfoDbData_1.2.11    \n",
       " [35] irlba_2.3.5.1               listenv_0.10.0             \n",
       " [37] spatstat.utils_3.2-0        seqLogo_1.68.0             \n",
       " [39] goftest_1.2-3               RSpectra_0.16-2            \n",
       " [41] spatstat.random_3.4-2       fitdistrplus_1.2-4         \n",
       " [43] parallelly_1.45.1           svglite_2.2.2              \n",
       " [45] RcppRoll_0.3.1              leiden_0.4.3.1             \n",
       " [47] codetools_0.2-20            DelayedArray_0.28.0        \n",
       " [49] shape_1.4.6.1               tidyselect_1.2.1           \n",
       " [51] farver_2.1.2                base64enc_0.1-3            \n",
       " [53] spatstat.explore_3.5-3      jsonlite_2.0.0             \n",
       " [55] GetoptLong_1.0.5            ellipsis_0.3.2             \n",
       " [57] progressr_0.18.0            iterators_1.0.14           \n",
       " [59] ggridges_0.5.7              survival_3.8-3             \n",
       " [61] systemfonts_1.3.1.9000      foreach_1.5.2              \n",
       " [63] tools_4.3.1                 ragg_1.5.0                 \n",
       " [65] TFMPvalue_0.0.9             ica_1.0-3                  \n",
       " [67] glue_1.8.0                  SparseArray_1.2.4          \n",
       " [69] IRdisplay_1.1               withr_3.0.2                \n",
       " [71] fastmap_1.2.0               rhdf5filters_1.14.1        \n",
       " [73] caTools_1.18.3              digest_0.6.37              \n",
       " [75] R6_2.6.1                    mime_0.13                  \n",
       " [77] colorspace_2.1-2            textshaping_0.3.7          \n",
       " [79] scattermore_1.2             tensor_1.5.1               \n",
       " [81] RSQLite_2.4.3               spatstat.data_3.1-9        \n",
       " [83] tidyr_1.3.1                 generics_0.1.4             \n",
       " [85] httr_1.4.7                  htmlwidgets_1.6.4          \n",
       " [87] S4Arrays_1.2.1              uwot_0.2.3                 \n",
       " [89] pkgconfig_2.0.3             blob_1.2.4                 \n",
       " [91] lmtest_0.9-40               XVector_0.42.0             \n",
       " [93] htmltools_0.5.8.1           profvis_0.4.0              \n",
       " [95] dotCall64_1.2               clue_0.3-66                \n",
       " [97] scales_1.4.0                png_0.1-8                  \n",
       " [99] spatstat.univar_3.1-4       rjson_0.2.23               \n",
       "[101] reshape2_1.4.4              uuid_1.2-1                 \n",
       "[103] nlme_3.1-168                GlobalOptions_0.1.2        \n",
       "[105] repr_1.1.7                  cachem_1.1.0               \n",
       "[107] zoo_1.8-14                  KernSmooth_2.23-26         \n",
       "[109] miniUI_0.1.2                pillar_1.11.1              \n",
       "[111] vctrs_0.6.5                 RANN_2.6.2                 \n",
       "[113] urlchecker_1.0.1            promises_1.5.0             \n",
       "[115] xtable_1.8-4                cluster_2.1.8.1            \n",
       "[117] evaluate_1.0.5              Rsamtools_2.18.0           \n",
       "[119] cli_3.6.5                   compiler_4.3.1             \n",
       "[121] rlang_1.1.6                 crayon_1.5.3               \n",
       "[123] future.apply_1.20.0         labeling_0.4.3             \n",
       "[125] fs_1.6.6                    stringi_1.8.7              \n",
       "[127] BiocParallel_1.36.0         viridisLite_0.4.2          \n",
       "[129] deldir_2.0-4                Biostrings_2.70.3          \n",
       "[131] lazyeval_0.2.2              spatstat.geom_3.6-0        \n",
       "[133] IRkernel_1.3.2              RcppHNSW_0.6.0             \n",
       "[135] bit64_4.6.0-1               future_1.67.0              \n",
       "[137] Rhdf5lib_1.24.2             shiny_1.11.1               \n",
       "[139] ROCR_1.0-11                 igraph_2.2.1               \n",
       "[141] memoise_2.0.1               fastmatch_1.1-6            \n",
       "[143] bit_4.6.0                  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
